src/include/sof/audio/eq_iir/iir.h

```diff
@@ -61,6 +62,50 @@ struct iir_state_df2t {
 	int64_t *delay; /* Pointer to IIR delay line */
 };
 
+/*
+ * \brief Returns the output of a biquad.
+ *
+ * 32 bit data, 32 bit coefficients and 64 bit state variables
+ * \param in input to the biquad Q1.31
+ * \param coef coefficients of the biquad Q2.30
+ * \param delay delay of the biquads Q3.61
+ */
+static inline int32_t iir_process_biquad(int32_t in, int32_t *coef,
+					 int64_t *delay)
+{
+	int32_t tmp;
+	int64_t acc;
+
+	/* Compute output: Delay is Q3.61
+	 * Q2.30 x Q1.31 -> Q3.61
+	 * Shift Q3.61 to Q3.31 with rounding
+	 */
+	acc = ((int64_t)coef[4]) * in + delay[0];
+	tmp = (int32_t)Q_SHIFT_RND(acc, 61, 31);
+
+	/* Compute first delay */
+	acc = delay[1];
+	acc += ((int64_t)coef[3]) * in; /* Coef  b1 */
+	acc += ((int64_t)coef[1]) * tmp; /* Coef a1 */
+	delay[0] = acc;
+
+	/* Compute second delay */
+	acc = ((int64_t)coef[2]) * in; /* Coef  b2 */
+	acc += ((int64_t)coef[0]) * tmp; /* Coef a2 */
+	delay[1] = acc;
+
+	/* Apply gain Q2.14 x Q1.31 -> Q3.45 */
+	acc = ((int64_t)coef[6]) * tmp; /* Gain */
+
+	/* Apply biquad output shift right parameter
+	 * simultaneously with Q3.45 to Q3.31 conversion. Then
+	 * saturate to 32 bits Q1.31 and prepare for next
+	 * biquad.
+	 */
+	acc = Q_SHIFT_RND(acc, 45 + coef[5], 31);
+	return sat_int32(acc);
+}
+
```

Comment:
moving this to a .c implementation? I am not sure it's good idea to define such a long inline function.

Reply:
The argument for that was that its a function that is called quite often in a tight loop and might suffer from constantly jumping on every sample.

Reply:
There would be performance hit for IIR EQ too with not inline, even more with typical higher order filters that here.

Reply:
Okay, sounds fair then. Hi @singalsu per my understanding, calling to this inline function in a tight loop will actually lead to high .text section consumption?

Reply:
I don't think the loop will be unrolled by compiler since the number of times is variable in every usage for it. It's true that this core is now duplicated. This crossover filter version does not use the SIMD version but once it will be updated (like IIR EQ) the code overhead will be much smaller. About every line above will then consume a single assembly instruction. The optimization can be be done by copying this part from IIR EQ, or by changing to generic Nth order IIRs when the optimization would be common for these two components.

Reply:
Also another advantage could be that including this snippet of code in a header file removes the dependency between IIR and Crossover at the Kconfig level.
i.e. COMP_CROSSOVER can be enabled without having to enable COMP_IIR in the Kconfig.

Reply:
Have you looked into optimizing code speed with IIR intrinsic version? The C version may reach near full load (MCPS) with higher output streams count.

In beamformer development I've removed inline of FIR core function. It has been already duplicated many times for EQ usage and beamformer would duplicate again. I don't think the performance hit will be large vs. code size hit. But IIR is a different case for core load vs. surrounding code load so I don't see issue in this if code structure still OK for optimized version. I've done pretty large changes afterwards to host the optimized versions and I seem to need to shuffle a bit things gain between FIR EQ and beamformer. Just my 2c :)

Reply:
I agree, we should be selecting the IIR function dynamically so that we can use a SIMD variant if available. This will be a followup PR due to timing constraints for @sebcarlucci

Reply:
Thanks, that sounds good.